File: backend/test/contract/aiCoach.contract.test.ts
import assert from 'node:assert/strict';
import test from 'node:test';
import { buildChatResponse } from '../../src/utils/chatResponse';

test('AI Coach response has a stable success and fallback envelope', () => {
  const draft = {
    type: 'EXPENSE' as const,
    amount: 25_000,
    categoryId: 'Food',
    walletId: 'wallet-1',
  };

  assert.deepEqual(buildChatResponse(' Periksa draft. ', draft), {
    reply: 'Periksa draft.',
    transactionDraft: draft,
    fallback: false,
  });
  assert.deepEqual(buildChatResponse('AI tidak tersedia.', null, true), {
    reply: 'AI tidak tersedia.',
    transactionDraft: null,
    fallback: true,
  });
  assert.equal(buildChatResponse('').reply, 'Respons Nala belum tersedia.');
});

File: mobile/test/contract/ai_coach_contract_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:nala/services/chat_service.dart';

void main() {
  test('AI Coach accepts the documented success and fallback envelopes', () {
    final success = parseChatReply({
      'reply': 'Periksa draft berikut.',
      'fallback': false,
      'transactionDraft': {
        'type': 'EXPENSE',
        'amount': 25000,
        'walletId': 'wallet-1',
        'categoryId': 'Food',
        'merchant': ' Kantin ',
        'unexpected': 'must not cross the boundary',
      },
    });
    final fallback = parseChatReply({
      'reply': 'Layanan AI sedang tidak tersedia.',
      'fallback': true,
      'transactionDraft': null,
    });

    expect(success?.message, 'Periksa draft berikut.');
    expect(success?.fallback, isFalse);
    expect(success?.transactionDraft, {
      'type': 'EXPENSE',
      'amount': 25000,
      'walletId': 'wallet-1',
      'categoryId': 'Food',
      'merchant': 'Kantin',
    });
    expect(fallback?.fallback, isTrue);
    expect(fallback?.transactionDraft, isNull);
  });

  test('AI Coach rejects malformed envelopes and unsafe drafts', () {
    expect(parseChatReply({'reply': '', 'fallback': false}), isNull);
    expect(parseChatReply({'reply': 'Halo', 'fallback': 'false'}), isNull);

    final response = parseChatReply({
      'reply': 'Draft tidak aman dibuang.',
      'fallback': false,
      'transactionDraft': {
        'type': 'EXPENSE',
        'amount': 25000,
        'walletId': 'wallet-1',
        'categoryId': 'SYSTEM_OVERRIDE',
      },
    });
    expect(response, isNotNull);
    expect(response?.transactionDraft, isNull);
  });
}
